airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/DetectStreamToFlush.java

```diff
@@ -0,0 +1,222 @@
+   * CDC where it's random).
+   * <p>
+   * This method also returns debug string with info that about the computation. We do it this way, so
+   * that the debug info that is printed is exactly what is used in the computation.
+   *
+   * @param stream stream
+   * @return is time triggered and a debug string
+   */
+  @VisibleForTesting
+  ImmutablePair<Boolean, String> isTimeTriggered(final StreamDescriptor stream) {
+    final Boolean isTimeTriggered = bufferDequeue.getTimeOfLastRecord(stream)
+        .map(time -> time.isBefore(Instant.now().minus(MAX_TIME_BETWEEN_REC_MIN, ChronoUnit.MINUTES)))
+        .orElse(false);
+
+    final String debugString = String.format("time trigger: %s", isTimeTriggered);
+
+    return ImmutablePair.of(isTimeTriggered, debugString);
+  }
+
+  /**
+   * For the size threshold, the size of the data in the queue is compared to the threshold that is
+   * passed into this method.
+   * <p>
+   * One caveat, is that if that stream already has a worker running, we "penalize" its size. We do
+   * this by computing what the size of the queue would be after the running workers for that queue
+   * complete. This is based on a dumb estimate of how much data a worker can process. There is an
+   * opportunity for optimization here, by being smarter about predicting how much data a running
+   * worker is likely to process.
+   * <p>
+   * This method also returns debug string with info that about the computation. We do it this way, so
+   * that the debug info that is printed is exactly what is used in the computation.
+   *
+   * @param stream stream
+   * @param queueSizeThresholdBytes min size threshold to determine if a queue is ready to flush
+   * @return is size triggered and a debug string
+   */
+  @VisibleForTesting
+  ImmutablePair<Boolean, String> isSizeTriggered(final StreamDescriptor stream, final long queueSizeThresholdBytes) {
+    final long currentQueueSize = bufferDequeue.getQueueSizeBytes(stream).orElseThrow();
+    final long sizeOfRunningWorkersEstimate = estimateSizeOfRunningWorkers(stream, currentQueueSize);
+    final long queueSizeAfterRunningWorkers = currentQueueSize - sizeOfRunningWorkersEstimate;
+    final boolean isSizeTriggered = queueSizeAfterRunningWorkers > queueSizeThresholdBytes;
+
+    final String debugString = String.format(
+        "size trigger: %s current threshold b: %s, queue size b: %s, penalty b: %s, after penalty b: %s",
+        isSizeTriggered,
+        AirbyteFileUtils.byteCountToDisplaySize(queueSizeThresholdBytes),
+        AirbyteFileUtils.byteCountToDisplaySize(currentQueueSize),
+        AirbyteFileUtils.byteCountToDisplaySize(sizeOfRunningWorkersEstimate),
+        AirbyteFileUtils.byteCountToDisplaySize(queueSizeAfterRunningWorkers));
+
+    return ImmutablePair.of(isSizeTriggered, debugString);
+  }
+
+  /**
+   * For a stream, determines how many bytes will be processed by CURRENTLY running workers. For the
```

Comment:
Is it possible for a stream to have more than one active worker? (I think so)

Reply:
yes!